src/lib/ReactCore_Internal.re

```diff
@@ -164,97 +164,139 @@ module Make = (Implementation: HostImplementation) => {
         | _ => assert false;
-      aux([], (list1, list2, list3))
+      aux([], (list1, list2, list3));
     };
     let map3 = (f, list1, list2, list3) =>
       rev_map3(f, List.rev(list1), List.rev(list2), List.rev(list3));
   };
 
   /*** Log of operations performed to update an instance tree. */
   module UpdateLog = {
+    type subtreeChange =
+      | Nested
+      | NoChange
+      | PrependElement(renderedElement)
+      | ReplaceElements(renderedElement, renderedElement);
+    type instanceUpdate =
+      | InstanceUpdate(
+                        instance('state, 'action, 'elementType),
+                        instance('state, 'action, 'elementType)
+                      ): instanceUpdate;
     type update = {
+      oldId: int,
+      newId: int,
+      instanceUpdate,
+      stateChanged: bool,
+      subTreeChanged: subtreeChange
+    };
+    type switchComponent = {
       oldId: int,
       newId: int,
       oldOpaqueInstance: opaqueInstance,
       newOpaqueInstance: opaqueInstance,
-      componentChanged: bool,
-      stateChanged: bool,
-      subTreeChanged: bool
+      subTreeChanged: subtreeChange
     };
     type entry =
       | UpdateInstance(update)
-      | NewRenderedElement(renderedElement);
+      | SwitchComponent(switchComponent);
     type t = ref(list(entry));
+    type topLevelUpdate = {
+      typ: subtreeChange,
+      updateLog: t
+    };
     let create = () => ref([]);
     let add = (updateLog, x) => updateLog := [x, ...updateLog^];
   };
   module Render = {
-    let createSelf = (~instance) : self(_) => {
-      state: instance.iState,
+    let createSelf = (~state, ~component, ~pendingStateUpdates) : self(_) => {
+      state,
       reduce: (payloadToAction, payload) => {
-        logString("Calling reduce on " ++ instance.component.debugName);
+        logString("Calling reduce on " ++ component.debugName);
         let action = payloadToAction(payload);
-        let stateUpdate = instance.component.reducer(action);
-        instance.pendingStateUpdates :=
-          [stateUpdate, ...instance.pendingStateUpdates^]
+        let stateUpdate = component.reducer(action);
+        pendingStateUpdates := [stateUpdate, ...pendingStateUpdates^];
       },
-      act: (action) => {
-        logString("Calling act on " ++ instance.component.debugName);
-        let stateUpdate = instance.component.reducer(action);
-        instance.pendingStateUpdates :=
-          [stateUpdate, ...instance.pendingStateUpdates^]
+      act: action => {
+        logString("Calling act on " ++ component.debugName);
+        let stateUpdate = component.reducer(action);
+        pendingStateUpdates := [stateUpdate, ...pendingStateUpdates^];
       }
     };
-    let createInstance = (~component, ~element, ~instanceSubTree, ~subElements) => {
-      let iState = component.initialState();
-      let id = GlobalState.instanceIdCounter^;
-      incr(GlobalState.instanceIdCounter);
+    let createInitialSelf = (~component) => {
```

Comment:
This should probably increment the global `id` counter and return the `id`. Now the mutation is in two places but always colocated to this call.